packages/compartment-mapper/src/types/compartment-map-schema.d.ts

```diff
@@ -0,0 +1,116 @@
+/**
+ * @file These types describe the schema of a `compartment-map.json`, which
+ * in turn describes how to load and link an application from storage, like a
+ * file system, web host, or zip archive.
+ */
+
+/* eslint-disable no-use-before-define */
+
+import type { SomePackagePolicy } from './policy-schema.js';
+import type { LiteralUnion } from './typescript.js';
+
+/**
+ * A compartment map describes how to construct an application as a graph of
+ * Compartments, each corresponding to Node.js style packaged modules.
+ */
+export type CompartmentMapDescriptor = {
+  tags: Array<string>;
+  entry: EntryDescriptor;
+  compartments: Record<string, CompartmentDescriptor>;
+};
+
+/**
+ * The entry descriptor of a compartment map denotes the root module of an
+ * application and the compartment that contains it.
+ */
+export type EntryDescriptor = {
+  compartment: string;
+  module: string;
+};
+
+/**
+ * A compartment descriptor corresponds to a single Compartment
+ * of an assembled Application and describes how to construct
+ * one for a given library or application `package.json`.
+ */
+export type CompartmentDescriptor = {
+  label: string;
+  /** shortest path of dependency names to this compartment */
+  path?: string[] | undefined;
+  /**
+   * the name of the originating package suitable for constructing a sourceURL
+   * prefix that will match it to files in a developer workspace.
+   */
+  name: string;
+  location: string;
+  /**
+   * whether this compartment was retained by any module in the solution. This
+   * property should never appear in an archived compartment map.
+   */
+  retained?: boolean | undefined;
+  modules: Record<string, ModuleDescriptor>;
+  scopes: Record<string, ScopeDescriptor>;
+  /** language for extension */
+  parsers: LanguageForExtension;
+  /** language for module specifier */
+  types: LanguageForModuleSpecifier;
+  /** policy specific to compartment */
+  policy: SomePackagePolicy;
+  /** List of compartment names this Compartment depends upon */
+  compartments: Set<string>;
+};
+
+/**
+ * For every module explicitly mentioned in an `exports` field of a
+ * `package.json`, there is a corresponding module descriptor.
+ */
+export type ModuleDescriptor = {
+  compartment?: string | undefined;
+  module?: string | undefined;
+  location?: string | undefined;
+  parser?: Language | undefined;
+  /** in base 16, hex */
+  sha512?: string | undefined;
+  exit?: string | undefined;
+  deferredError?: string | undefined;
+  retained?: boolean | undefined;
```

Comment:
Nit: tsc generated redundant `| undefined` here and I missed them in my first pass. Circle back.